pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,528 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"path/filepath"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	utilstore "k8s.io/kubernetes/pkg/kubelet/util/store"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// bitsDataElement is the number of bits in a bitArray.data element.
+const bitsDataElement = 32
+
+type bitArray struct {
+	data       []uint32
+	firstIndex int
+}
+
+func makeBitArray(size uint32) *bitArray {
+	m := bitArray{
+		data:       make([]uint32, (size+bitsDataElement-1)/bitsDataElement),
+		firstIndex: 0,
+	}
+	return &m
+}
+
+func (b *bitArray) set(index uint32) {
+	b.data[index/bitsDataElement] |= (uint32(1) << (index % bitsDataElement))
+}
+
+func (b *bitArray) isSet(index uint32) bool {
+	return (b.data[index/bitsDataElement]>>(index%bitsDataElement))&0x1 == 1
+}
+
+func (b *bitArray) findAvailable() (uint32, bool) {
+	for i := b.firstIndex; i < len(b.data); i++ {
+		// Check if all bits are used (all 1s).
+		if b.data[i] == math.MaxUint32 {
+			continue
+		}
+		for j := uint32(0); j < bitsDataElement; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsDataElement + j
+				b.set(v)
+				// Update firstIndex to the current
+				// data element since there are no other
+				// unset bits before the current index.
+				b.firstIndex = int(i)
+				return v, true
+			}
+		}
+	}
+	return 0, false
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsDataElement
+	// update firstIndex if the index found is less than
+	// the current one.
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	// clear the bit by ANDing the data element with the
+	// complement of the bitmask to be cleared.
+	b.data[i] &= ^(1 << (index % bitsDataElement))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = (1 << 16)
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
+
+// Create a new map when we removed enough pods to avoid memory leaks
+// since Go maps never free memory.
+const mapReInitializeThreshold = 1000
+
+type userNsPodsManager interface {
+	getPodDir(podUID types.UID) string
+	listPodsFromDisk() ([]types.UID, error)
+}
+
+type usernsManager struct {
+	used         *bitArray
+	usedBy       map[string]uint32
```

Comment:
Can the key type be a `types.UID` here to make it more clear that this is a pod UID? I'd also add a comment to that effect, and update the associated method signatures.

Reply:
Done